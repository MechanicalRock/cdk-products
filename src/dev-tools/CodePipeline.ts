import { Pipeline, Stage } from '@aws-cdk/aws-codepipeline';
import { PipelineBuildAction, PipelineProject, CodeCommitSource, ComputeType, LinuxBuildImage } from '@aws-cdk/aws-codebuild';
import { PipelineSourceAction, Repository } from '@aws-cdk/aws-codecommit';
import { Stack } from '@aws-cdk/cdk';
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';

export interface CodePipelineProps {
    stack: Stack,
    logicalName: string,
    pipelineName: string,
    codeRepository: Repository
}

export class CodePipeline {
    pipeline: Pipeline;
    constructor(private props: CodePipelineProps) {
        this.create();
    }

    create = () => {
        // Create the pipeline
        const pipeline = new Pipeline(this.props.stack, this.props.logicalName, {
            pipelineName: this.props.pipelineName,
            restartExecutionOnUpdate: true
        });
        this.pipeline = pipeline;

        const codeBuildRole = this.createCodeBuildRole();
        this.createSourceStage();
        this.createDevStage(codeBuildRole);
        this.createProdStage(codeBuildRole);
    }

    private createCodeBuildRole = (): Role => {
        const codeBuildProjectRole = new Role(this.props.stack, 'CodeBuildRole', {
            assumedBy: new ServicePrincipal('codebuild.amazonaws.com')
        });
        codeBuildProjectRole.attachManagedPolicy('arn:aws:iam::aws:policy/AdministratorAccess');
        return codeBuildProjectRole;
    }

    private createProdStage = (role: Role) => {
        // Prod stage and build & test action
        const prodStage = new Stage(this.props.stack, 'ProdStage', {
            pipeline: this.pipeline
        });

        const prodCodeBuildProject = this.createCodeBuildProject(role, 'DeployToProd');

        new PipelineBuildAction(this.props.stack, 'prodDeployAction', {
            project: prodCodeBuildProject,
            stage: prodStage
        })
    }

    private createDevStage = (role: Role) => {
        // Dev stage and build & test action
        const devStage = new Stage(this.props.stack, 'DevStage', {
            pipeline: this.pipeline
        });

        const buildAndTestCodeBuildProject = this.createCodeBuildProject(role, 'BuildAndTest');

        new PipelineBuildAction(this.props.stack, 'BuildAndTestAction', {
            project: buildAndTestCodeBuildProject,
            stage: devStage
        })
    }

    private createCodeBuildProject = (role: Role, logicalName: string): PipelineProject => {
        return new PipelineProject(this.props.stack, logicalName, {
            environment: {
                buildImage: LinuxBuildImage.UBUNTU_14_04_NODEJS_8_11_0
            },
            role
        })
    }

    private createSourceStage = () => {
        // Source stage and action
        const sourceStage = new Stage(this.props.stack, 'SourceStage', {
            pipeline: this.pipeline,
        });

        new PipelineSourceAction(this.props.stack, 'SourceAction', {
            repository: this.props.codeRepository,
            stage: sourceStage,
            pollForSourceChanges: true
        })
    }
}