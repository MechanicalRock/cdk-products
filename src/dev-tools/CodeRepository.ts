import { Repository } from '@aws-cdk/aws-codecommit';
import { Stack, Output } from '@aws-cdk/cdk';

export class CodeRepository {
    repository: Repository;

    constructor(private stack: Stack, private repoLogicalName: string, private repoName: string) {
        this.repository = this.create();
        this.createOutputs();
    }

    create(): Repository {
        return new Repository(this.stack, this.repoLogicalName, {
            repositoryName: this.repoName,
        })
    }

    createOutputs = () => {
        new Output(this.stack, 'RepositoryHttpUrl', {
            value: this.repository.repositoryCloneUrlHttp,
            disableExport: true
        })
    }
}