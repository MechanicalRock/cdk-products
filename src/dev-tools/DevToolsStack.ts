import { Stack, App } from '@aws-cdk/cdk';
import { CodeRepository } from './CodeRepository';
import { CodePipeline } from './CodePipeline';

export class DevToolsStack extends Stack {

    constructor(app: App, stackName: string) {
        super(app, stackName);
        this.create();
    }

    create = () => {
        // Create a code commit repo
        const repository = new CodeRepository(this, 'SampleStaticSiteRepo', 'sample-static-site').repository;

        // Create a code pipeline
        new CodePipeline({
            stack: this,
            logicalName: 'CodePipeline',
            pipelineName: 'code-pipeline',
            codeRepository: repository
        });
    }
}