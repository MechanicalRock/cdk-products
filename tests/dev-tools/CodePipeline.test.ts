import { Stack } from '@aws-cdk/cdk';
import * as yaml from '../utils/Yaml';
import { CodePipeline } from '../../src/dev-tools/CodePipeline';
import { Repository } from '@aws-cdk/aws-codecommit';

describe('CodePipeline', () => {
    it('should match the template snapshot', () => {
        // Given
        const stack = new Stack();
        const repository = new Repository(stack, 'Repository', {
            repositoryName: 'code-repo'
        })

        // When
        new CodePipeline({
            stack,
            logicalName: 'CodePipeline',
            pipelineName: 'code-pipeline',
            codeRepository: repository
        });

        // Then
        expect(yaml.toYamlString(stack.toCloudFormation())).toMatchSnapshot();
    })
})