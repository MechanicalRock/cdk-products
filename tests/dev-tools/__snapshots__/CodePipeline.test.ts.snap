// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodePipeline should match the template snapshot 1`] = `
"Resources:
  Repository22E53BBD:
    Type: 'AWS::CodeCommit::Repository'
    Properties:
      RepositoryName: code-repo
      Triggers: []
  CodePipelineArtifactsBucketF1E925CF:
    Type: 'AWS::S3::Bucket'
    DeletionPolicy: Retain
  CodePipelineRoleB3A660B4:
    Type: 'AWS::IAM::Role'
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Action: 'sts:AssumeRole'
            Effect: Allow
            Principal:
              Service: codepipeline.amazonaws.com
        Version: '2012-10-17'
  CodePipelineRoleDefaultPolicy8D520A8D:
    Type: 'AWS::IAM::Policy'
    Properties:
      PolicyDocument:
        Statement:
          - Action:
              - 's3:GetObject*'
              - 's3:GetBucket*'
              - 's3:List*'
              - 's3:DeleteObject*'
              - 's3:PutObject*'
              - 's3:Abort*'
            Effect: Allow
            Resource:
              - 'Fn::GetAtt':
                  - CodePipelineArtifactsBucketF1E925CF
                  - Arn
              - 'Fn::Join':
                  - ''
                  - - 'Fn::GetAtt':
                        - CodePipelineArtifactsBucketF1E925CF
                        - Arn
                    - /*
          - Action:
              - 'codecommit:GetBranch'
              - 'codecommit:GetCommit'
              - 'codecommit:UploadArchive'
              - 'codecommit:GetUploadArchiveStatus'
              - 'codecommit:CancelUploadArchive'
            Effect: Allow
            Resource:
              'Fn::GetAtt':
                - Repository22E53BBD
                - Arn
          - Action:
              - 'codebuild:BatchGetBuilds'
              - 'codebuild:StartBuild'
              - 'codebuild:StopBuild'
            Effect: Allow
            Resource:
              'Fn::GetAtt':
                - BuildAndTestB9A2F419
                - Arn
          - Action:
              - 'codebuild:BatchGetBuilds'
              - 'codebuild:StartBuild'
              - 'codebuild:StopBuild'
            Effect: Allow
            Resource:
              'Fn::GetAtt':
                - DeployToProd5A783BE8
                - Arn
        Version: '2012-10-17'
      PolicyName: CodePipelineRoleDefaultPolicy8D520A8D
      Roles:
        - Ref: CodePipelineRoleB3A660B4
  CodePipelineB74E5936:
    Type: 'AWS::CodePipeline::Pipeline'
    Properties:
      ArtifactStore:
        Location:
          Ref: CodePipelineArtifactsBucketF1E925CF
        Type: S3
      RoleArn:
        'Fn::GetAtt':
          - CodePipelineRoleB3A660B4
          - Arn
      Stages:
        - Actions:
            - ActionTypeId:
                Category: Source
                Owner: AWS
                Provider: CodeCommit
                Version: '1'
              Configuration:
                RepositoryName:
                  'Fn::GetAtt':
                    - Repository22E53BBD
                    - Name
                BranchName: master
                PollForSourceChanges: true
              InputArtifacts: []
              Name: SourceAction
              OutputArtifacts:
                - Name: Artifact_SourceAction
              RunOrder: 1
          Name: SourceStage
        - Actions:
            - ActionTypeId:
                Category: Build
                Owner: AWS
                Provider: CodeBuild
                Version: '1'
              Configuration:
                ProjectName:
                  Ref: BuildAndTestB9A2F419
              InputArtifacts:
                - Name: Artifact_SourceAction
              Name: BuildAndTestAction
              OutputArtifacts:
                - Name: Artifact_BuildAndTestAction
              RunOrder: 1
          Name: DevStage
        - Actions:
            - ActionTypeId:
                Category: Build
                Owner: AWS
                Provider: CodeBuild
                Version: '1'
              Configuration:
                ProjectName:
                  Ref: DeployToProd5A783BE8
              InputArtifacts:
                - Name: Artifact_BuildAndTestAction
              Name: prodDeployAction
              OutputArtifacts:
                - Name: Artifact_prodDeployAction
              RunOrder: 1
          Name: ProdStage
      Name: code-pipeline
      RestartExecutionOnUpdate: true
    DependsOn:
      - CodePipelineRoleB3A660B4
      - CodePipelineRoleDefaultPolicy8D520A8D
  CodeBuildRole728CBADE:
    Type: 'AWS::IAM::Role'
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Action: 'sts:AssumeRole'
            Effect: Allow
            Principal:
              Service: codebuild.amazonaws.com
        Version: '2012-10-17'
      ManagedPolicyArns:
        - 'arn:aws:iam::aws:policy/AdministratorAccess'
  CodeBuildRoleDefaultPolicy829527DE:
    Type: 'AWS::IAM::Policy'
    Properties:
      PolicyDocument:
        Statement:
          - Action:
              - 'logs:CreateLogGroup'
              - 'logs:CreateLogStream'
              - 'logs:PutLogEvents'
            Effect: Allow
            Resource:
              - 'Fn::Join':
                  - ''
                  - - 'arn:'
                    - Ref: 'AWS::Partition'
                    - ':logs:'
                    - Ref: 'AWS::Region'
                    - ':'
                    - Ref: 'AWS::AccountId'
                    - ':log-group:/aws/codebuild/'
                    - Ref: BuildAndTestB9A2F419
              - 'Fn::Join':
                  - ''
                  - - 'arn:'
                    - Ref: 'AWS::Partition'
                    - ':logs:'
                    - Ref: 'AWS::Region'
                    - ':'
                    - Ref: 'AWS::AccountId'
                    - ':log-group:/aws/codebuild/'
                    - Ref: BuildAndTestB9A2F419
                    - ':*'
          - Action:
              - 's3:GetObject*'
              - 's3:GetBucket*'
              - 's3:List*'
              - 's3:DeleteObject*'
              - 's3:PutObject*'
              - 's3:Abort*'
            Effect: Allow
            Resource:
              - 'Fn::GetAtt':
                  - CodePipelineArtifactsBucketF1E925CF
                  - Arn
              - 'Fn::Join':
                  - ''
                  - - 'Fn::GetAtt':
                        - CodePipelineArtifactsBucketF1E925CF
                        - Arn
                    - /*
          - Action:
              - 'logs:CreateLogGroup'
              - 'logs:CreateLogStream'
              - 'logs:PutLogEvents'
            Effect: Allow
            Resource:
              - 'Fn::Join':
                  - ''
                  - - 'arn:'
                    - Ref: 'AWS::Partition'
                    - ':logs:'
                    - Ref: 'AWS::Region'
                    - ':'
                    - Ref: 'AWS::AccountId'
                    - ':log-group:/aws/codebuild/'
                    - Ref: DeployToProd5A783BE8
              - 'Fn::Join':
                  - ''
                  - - 'arn:'
                    - Ref: 'AWS::Partition'
                    - ':logs:'
                    - Ref: 'AWS::Region'
                    - ':'
                    - Ref: 'AWS::AccountId'
                    - ':log-group:/aws/codebuild/'
                    - Ref: DeployToProd5A783BE8
                    - ':*'
          - Action:
              - 's3:GetObject*'
              - 's3:GetBucket*'
              - 's3:List*'
              - 's3:DeleteObject*'
              - 's3:PutObject*'
              - 's3:Abort*'
            Effect: Allow
            Resource:
              - 'Fn::GetAtt':
                  - CodePipelineArtifactsBucketF1E925CF
                  - Arn
              - 'Fn::Join':
                  - ''
                  - - 'Fn::GetAtt':
                        - CodePipelineArtifactsBucketF1E925CF
                        - Arn
                    - /*
        Version: '2012-10-17'
      PolicyName: CodeBuildRoleDefaultPolicy829527DE
      Roles:
        - Ref: CodeBuildRole728CBADE
  BuildAndTestB9A2F419:
    Type: 'AWS::CodeBuild::Project'
    Properties:
      Artifacts:
        Type: CODEPIPELINE
      Environment:
        ComputeType: BUILD_GENERAL1_SMALL
        Image: 'aws/codebuild/nodejs:8.11.0'
        PrivilegedMode: false
        Type: LINUX_CONTAINER
      ServiceRole:
        'Fn::GetAtt':
          - CodeBuildRole728CBADE
          - Arn
      Source:
        Type: CODEPIPELINE
  DeployToProd5A783BE8:
    Type: 'AWS::CodeBuild::Project'
    Properties:
      Artifacts:
        Type: CODEPIPELINE
      Environment:
        ComputeType: BUILD_GENERAL1_SMALL
        Image: 'aws/codebuild/nodejs:8.11.0'
        PrivilegedMode: false
        Type: LINUX_CONTAINER
      ServiceRole:
        'Fn::GetAtt':
          - CodeBuildRole728CBADE
          - Arn
      Source:
        Type: CODEPIPELINE
"
`;
